In [1]:
import pandas as pd
import numpy as np

2. Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

- In the context of this problem, what is a false positive?
- In the context of this problem, what is a false negative?
- How would you describe this model?

In [2]:
total = 46 + 7 + 13 + 34

In [43]:
# accuracy = (TP + TN) / (TP + FP + FN + TN)
accuracy = (46 + 34) / total
print(f'the accuracy is {accuracy:.2%}')

the accuracy is 80.00%


In [44]:
# recall = TP / (TP + FN)
recall = 46 / (46 + 7)
print(f'the recall is {recall:.2%}')

the recall is 86.79%


In [45]:
# precision = TP / (TP + FP)
precision = 46 / (46 + 13)
print(f'the precision is {precision:.2%}')

the precision is 77.97%


In the context of this problem, what is a false positive?
- predicting dog when it is actually a cat

In the context of this problem, what is a false negative?
- predicting cat when it is actually a dog

How would you describe this model?
- around 80% correct, 86.8% chance to predict dog out of all dogs, 78% chance to be correct in predicting dog.
- model is best at predicting dog when it is actually a dog

3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


In [6]:
c3 = pd.read_csv('c3.csv')

In [7]:
c3.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

An internal team wants to investigate the cause of the manufacturing defects. 
They tell you that they want to identify as many of the ducks 
that have a defect as possible. Which evaluation metric would be appropriate here? 
Which model would be the best fit for this use case?

In [8]:
# Which evaluation metric would be appropriate here? 
# we would want to look at the model Recall
recall_subset = c3[c3.actual == 'Defect']
model1_recall = (recall_subset.actual == recall_subset.model1).mean()
model2_recall = (recall_subset.actual == recall_subset.model2).mean()
model3_recall = (recall_subset.actual == recall_subset.model3).mean()

In [9]:
print(f'model 1 recall is: {model1_recall:.2%}')
print(f'model 2 recall is: {model2_recall:.2%}')
print(f'model 3 recall is: {model3_recall:.2%}')

model 1 recall is: 50.00%
model 2 recall is: 56.25%
model 3 recall is: 81.25%


Which model would be the best fit for this use case?

Model 3 has the best recall so is the best for this use case

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [11]:
# Which evaluation metric would be appropriate here?

# This is a poorly worded problem!! 
# The manufacturer would absolutely only care about
# catching every defective duck in this scenario regardless of false positives, 
# which is Recall, but the question intent is asking for precision.

# we will want to use the model Precision
model1_subset = c3[c3.model1 == 'Defect']
model1_precision = (model1_subset.actual == model1_subset.model1).mean()
model2_subset = c3[c3.model2 == 'Defect']
model2_precision = (model2_subset.actual == model2_subset.model2).mean()
model3_subset = c3[c3.model3 == 'Defect']
model3_precision = (model3_subset.actual == model3_subset.model3).mean()

In [12]:
print(f'model 1 precision is: {model1_precision:.2%}')
print(f'model 2 precision is: {model2_precision:.2%}')
print(f'model 3 precision is: {model3_precision:.2%}')

model 1 precision is: 80.00%
model 2 precision is: 10.00%
model 3 precision is: 13.13%


Which model would be the best fit for this use case?

model 1 is the best for this use case

4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

- In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
- Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?
- Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?

In [14]:
gyp = pd.read_csv('gives_you_paws.csv')
gyp.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [15]:
gyp.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [16]:
gyp['baseline'] = 'dog'

In [17]:
gyp.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [18]:
model1_accuracy = (gyp.actual == gyp.model1).mean()
model2_accuracy = (gyp.actual == gyp.model2).mean()
model3_accuracy = (gyp.actual == gyp.model3).mean()
model4_accuracy = (gyp.actual == gyp.model4).mean()
baseline_accuracy = (gyp.actual == gyp.baseline).mean()

In [19]:
print(f'baseline model accuracy is: {baseline_accuracy:.2%}')
print(f'model 1 accuracy is: {model1_accuracy:.2%}')
print(f'model 2 accuracy is: {model2_accuracy:.2%}')
print(f'model 3 accuracy is: {model3_accuracy:.2%}')
print(f'model 4 accuracy is: {model4_accuracy:.2%}')

baseline model accuracy is: 65.08%
model 1 accuracy is: 80.74%
model 2 accuracy is: 63.04%
model 3 accuracy is: 50.96%
model 4 accuracy is: 74.26%


In terms of accuracy, how do the various models compare to the baseline model? 
- Are any of the models better than the baseline?

model 1 and model 4 are better than the baseline accuracy

model 1 is the best accuracy for this use case

In [21]:
# Suppose you are working on a team that solely deals with dog pictures. 
# Which of these models would you recommend?

# another poorly worded problem, 
# are we looking to get as many dog pictures as possible?
# or do we want to make sure we only want dog pictures since the team only deals with dogs?

# I interpreted it as we want to be correct in predicting dog
# for this use case we would want to look at model precision, since we only want dog pics

In [22]:
baseline_subset = gyp[gyp.baseline == 'dog']
baseline_precision = (baseline_subset.actual == baseline_subset.baseline).mean()
model1_subset = gyp[gyp.model1 == 'dog']
model1_precision = (model1_subset.actual == model1_subset.model1).mean()
model2_subset = gyp[gyp.model2 == 'dog']
model2_precision = (model2_subset.actual == model2_subset.model2).mean()
model3_subset = gyp[gyp.model3 == 'dog']
model3_precision = (model3_subset.actual == model3_subset.model3).mean()
model4_subset = gyp[gyp.model4 == 'dog']
model4_precision = (model4_subset.actual == model4_subset.model4).mean()

In [67]:
print(f'baseline model precision is: {baseline_precision:.2%}')
print(f'model 1 precision is: {model1_precision:.2%}')
print(f'model 2 precision is: {model2_precision:.2%}')
print(f'model 3 precision is: {model3_precision:.2%}')
print(f'model 4 precision is: {model4_precision:.2%}')

baseline model precision is: 65.08%
model 1 precision is: 89.00%
model 2 precision is: 89.32%
model 3 precision is: 65.99%
model 4 precision is: 73.12%


Which of these models would you recommend?

model 2 precision is the best for this use case

In [25]:
# Suppose you are working on a team that solely deals with cat pictures. 
# Which of these models would you recommend?

# I understood the question to be that we want to minimize dog pictures
# we will look for model negative predictive value

In [26]:
baseline_subset = gyp[gyp.baseline == 'cat']
baseline_npv = (baseline_subset.actual == baseline_subset.baseline).mean()
model1_subset = gyp[gyp.model1 == 'cat']
model1_npv = (model1_subset.actual == model1_subset.model1).mean()
model2_subset = gyp[gyp.model2 == 'cat']
model2_npv = (model2_subset.actual == model2_subset.model2).mean()
model3_subset = gyp[gyp.model3 == 'cat']
model3_npv = (model3_subset.actual == model3_subset.model3).mean()
model4_subset = gyp[gyp.model4 == 'cat']
model4_npv = (model4_subset.actual == model4_subset.model4).mean()

In [27]:
print(f'baseline model negative predictive value is: {baseline_npv:.2%}')
print(f'model 1 negative predictive value is: {model1_npv:.2%}')
print(f'model 2 negative predictive value is: {model2_npv:.2%}')
print(f'model 3 negative predictive value is: {model3_npv:.2%}')
print(f'model 4 negative predictive value is: {model4_npv:.2%}')

baseline model negative predictive value is: nan%
model 1 negative predictive value is: 68.98%
model 2 negative predictive value is: 48.41%
model 3 negative predictive value is: 35.83%
model 4 negative predictive value is: 80.72%


Which of these models would you recommend?

model 4 negative predictive value is the best for this use case

In [29]:
from sklearn import metrics

5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- sklearn.metrics.accuracy_score
- sklearn.metrics.precision_score
- sklearn.metrics.recall_score
- sklearn.metrics.classification_report

In [30]:
# redoing c3 recall
c3_model1_recall = metrics.recall_score(c3.actual, c3.model1, pos_label='Defect')
c3_model2_recall = metrics.recall_score(c3.actual, c3.model2, pos_label='Defect')
c3_model3_recall = metrics.recall_score(c3.actual, c3.model3, pos_label='Defect')

In [31]:
print(f'c3 model 1 recall using sklearn.metrics is: {c3_model1_recall:.2%}')
print(f'c3 model 2 recall using sklearn.metrics is: {c3_model2_recall:.2%}')
print(f'c3 model 3 recall using sklearn.metrics is: {c3_model3_recall:.2%}')

c3 model 1 recall using sklearn.metrics is: 50.00%
c3 model 2 recall using sklearn.metrics is: 56.25%
c3 model 3 recall using sklearn.metrics is: 81.25%


In [32]:
# c3 precision using sklearn
c3_model1_precision = metrics.precision_score(c3.actual, c3.model1, pos_label='Defect')
c3_model2_precision = metrics.precision_score(c3.actual, c3.model2, pos_label='Defect')
c3_model3_precision = metrics.precision_score(c3.actual, c3.model3, pos_label='Defect')

In [33]:
print(f'c3 model 1 precision using sklearn.metrics is: {c3_model1_precision:.2%}')
print(f'c3 model 2 precision using sklearn.metrics is: {c3_model2_precision:.2%}')
print(f'c3 model 3 precision using sklearn.metrics is: {c3_model3_precision:.2%}')

c3 model 1 precision using sklearn.metrics is: 80.00%
c3 model 2 precision using sklearn.metrics is: 10.00%
c3 model 3 precision using sklearn.metrics is: 13.13%


In [34]:
gyp_base_accuracy = metrics.accuracy_score(gyp.actual, gyp.baseline)
gyp_model1_accuracy = metrics.accuracy_score(gyp.actual, gyp.model1)
gyp_model2_accuracy = metrics.accuracy_score(gyp.actual, gyp.model2)
gyp_model3_accuracy = metrics.accuracy_score(gyp.actual, gyp.model3)
gyp_model4_accuracy = metrics.accuracy_score(gyp.actual, gyp.model4)

In [35]:
print(f'gyp baseline accuracy using sklearn.metrics is: {gyp_base_accuracy:.2%}')
print(f'gyp model 1 accuracy using sklearn.metrics is: {gyp_model1_accuracy:.2%}')
print(f'gyp model 2 accuracy using sklearn.metrics is: {gyp_model2_accuracy:.2%}')
print(f'gyp model 3 accuracy using sklearn.metrics is: {gyp_model3_accuracy:.2%}')
print(f'gyp model 4 accuracy using sklearn.metrics is: {gyp_model4_accuracy:.2%}')

gyp baseline accuracy using sklearn.metrics is: 65.08%
gyp model 1 accuracy using sklearn.metrics is: 80.74%
gyp model 2 accuracy using sklearn.metrics is: 63.04%
gyp model 3 accuracy using sklearn.metrics is: 50.96%
gyp model 4 accuracy using sklearn.metrics is: 74.26%


In [36]:
gyp_base_precision = metrics.precision_score(gyp.actual, gyp.baseline, pos_label='dog')
gyp_model1_precision = metrics.precision_score(gyp.actual, gyp.model1, pos_label='dog')
gyp_model2_precision = metrics.precision_score(gyp.actual, gyp.model2, pos_label='dog')
gyp_model3_precision = metrics.precision_score(gyp.actual, gyp.model3, pos_label='dog')
gyp_model4_precision = metrics.precision_score(gyp.actual, gyp.model4, pos_label='dog')

In [37]:
print(f'gyp baseline precision using sklearn.metrics is: {gyp_base_precision:.2%}')
print(f'gyp model 1 precision using sklearn.metrics is: {gyp_model1_precision:.2%}')
print(f'gyp model 2 precision using sklearn.metrics is: {gyp_model2_precision:.2%}')
print(f'gyp model 3 precision using sklearn.metrics is: {gyp_model3_precision:.2%}')
print(f'gyp model 4 precision using sklearn.metrics is: {gyp_model4_precision:.2%}')

gyp baseline precision using sklearn.metrics is: 65.08%
gyp model 1 precision using sklearn.metrics is: 89.00%
gyp model 2 precision using sklearn.metrics is: 89.32%
gyp model 3 precision using sklearn.metrics is: 65.99%
gyp model 4 precision using sklearn.metrics is: 73.12%


In [65]:
gyp_base_npv = metrics.precision_score(gyp.actual, gyp.baseline, 
                                       pos_label='cat',
                                       zero_division=0)
gyp_model1_npv = metrics.precision_score(gyp.actual, gyp.model1, pos_label='cat')
gyp_model2_npv = metrics.precision_score(gyp.actual, gyp.model2, pos_label='cat')
gyp_model3_npv = metrics.precision_score(gyp.actual, gyp.model3, pos_label='cat')
gyp_model4_npv = metrics.precision_score(gyp.actual, gyp.model4, pos_label='cat')

In [66]:
print(f'gyp baseline negative predictive value using\
 sklearn.metrics is: {gyp_base_npv:.2%}')
print(f'gyp model 1 negative predictive value using\
 sklearn.metrics is: {gyp_model1_npv:.2%}')
print(f'gyp model 2 negative predictive value using\
 sklearn.metrics is: {gyp_model2_npv:.2%}')
print(f'gyp model 3 negative predictive value using\
 sklearn.metrics is: {gyp_model3_npv:.2%}')
print(f'gyp model 4 negative predictive value using\
 sklearn.metrics is: {gyp_model4_npv:.2%}')

gyp baseline negative predictive value using sklearn.metrics is: 0.00%
gyp model 1 negative predictive value using sklearn.metrics is: 68.98%
gyp model 2 negative predictive value using sklearn.metrics is: 48.41%
gyp model 3 negative predictive value using sklearn.metrics is: 35.83%
gyp model 4 negative predictive value using sklearn.metrics is: 80.72%


In [69]:
print('baseline stats:')
print(metrics.classification_report(gyp.actual, gyp.baseline, 
                                    target_names=['cat', 'dog'],
                                    zero_division=0))
print('-----------------------------------------------------')
print('model 1 stats:')
print(metrics.classification_report(gyp.actual, gyp.model1, 
                                    target_names=['cat', 'dog']))
print('-----------------------------------------------------')
print('model 2 stats:')
print(metrics.classification_report(gyp.actual, gyp.model2, 
                                    target_names=['cat', 'dog']))
print('-----------------------------------------------------')
print('model 3 stats:')
print(metrics.classification_report(gyp.actual, gyp.model3, 
                                    target_names=['cat', 'dog']))
print('-----------------------------------------------------')
print('model 4 stats:')
print(metrics.classification_report(gyp.actual, gyp.model4, 
                                    target_names=['cat', 'dog']))

baseline stats:
              precision    recall  f1-score   support

         cat       0.00      0.00      0.00      1746
         dog       0.65      1.00      0.79      3254

    accuracy                           0.65      5000
   macro avg       0.33      0.50      0.39      5000
weighted avg       0.42      0.65      0.51      5000

-----------------------------------------------------
model 1 stats:
              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

-----------------------------------------------------
model 2 stats:
              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                

In [78]:
for col in c3:
    if col != 'actual':
        print(f'---------------------{col}---------------------------')
        print(f'{col} stats:')
        print(metrics.classification_report(c3.actual, c3[col], 
                                            target_names=['Defect', 'No Defect'],
                                            zero_division=0))

---------------------model1---------------------------
model1 stats:
              precision    recall  f1-score   support

      Defect       0.80      0.50      0.62        16
   No Defect       0.96      0.99      0.97       184

    accuracy                           0.95       200
   macro avg       0.88      0.74      0.79       200
weighted avg       0.95      0.95      0.94       200

---------------------model2---------------------------
model2 stats:
              precision    recall  f1-score   support

      Defect       0.10      0.56      0.17        16
   No Defect       0.94      0.56      0.70       184

    accuracy                           0.56       200
   macro avg       0.52      0.56      0.44       200
weighted avg       0.87      0.56      0.66       200

---------------------model3---------------------------
model3 stats:
              precision    recall  f1-score   support

      Defect       0.13      0.81      0.23        16
   No Defect       0.97      0

In [75]:
for col in gyp:
    if col != 'actual':
        print(f'---------------------{col}---------------------------')
        print(f'{col} stats:')
        print(metrics.classification_report(gyp.actual, gyp[col], 
                                            target_names=['cat', 'dog'],
                                            zero_division=0))

---------------------model1---------------------------
model1 stats:
              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

---------------------model2---------------------------
model2 stats:
              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000

---------------------model3---------------------------
model3 stats:
              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0